common/src/gas_provider/node.rs

```diff
@@ -70,6 +75,72 @@ impl<T> From<ReservationId> for GasNodeId<T, ReservationId> {
     }
 }
 
+#[derive(Clone, Copy, Decode, Encode, Debug, Default, PartialEq, Eq, TypeInfo)]
+#[codec(crate = codec)]
+#[scale_info(crate = scale_info)]
+pub struct NodeLock<Balance>([Balance; 4]);
+
+impl<'a, Balance> NodeLock<Balance> {
+    pub fn iter(&'a self) -> Iter<'a, Balance> {
+        self.0.iter()
+    }
+}
+
+impl<Balance> Index<LockIdentifier> for NodeLock<Balance> {
+    type Output = Balance;
+
+    fn index(&self, index: LockIdentifier) -> &Self::Output {
+        match index {
+            LockIdentifier::Mailbox => &self.0[0],
+            LockIdentifier::Waitlist => &self.0[1],
+            LockIdentifier::Reservation => &self.0[2],
+            LockIdentifier::DispatchStash => &self.0[3],
+        }
+    }
+}
+
+impl<Balance> IndexMut<LockIdentifier> for NodeLock<Balance> {
+    fn index_mut(&mut self, index: LockIdentifier) -> &mut Self::Output {
+        match index {
+            LockIdentifier::Mailbox => &mut self.0[0],
+            LockIdentifier::Waitlist => &mut self.0[1],
+            LockIdentifier::Reservation => &mut self.0[2],
+            LockIdentifier::DispatchStash => &mut self.0[3],
+        }
+    }
+}
+
+impl<Balance: Zero + Copy> Zero for NodeLock<Balance> {
+    fn zero() -> Self {
+        Self([Balance::zero(); 4])
+    }
+
+    fn set_zero(&mut self) {
+        self.0.iter_mut().for_each(|x| x.set_zero());
+    }
+
+    fn is_zero(&self) -> bool {
+        self.0.iter().all(|x| x.is_zero())
+    }
+}
+
+impl<Balance: Add + Zero + Copy> Add<Self> for NodeLock<Balance> {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self::Output {
```

Comment:
We don't use this function anywhere, it's just a requirement that a `Zero` type has to be `Add`, hence such inefficient implementation. Anyway, yours looks good - will go with that one

Reply:
> Furthermore, I'd go with saturating_add naming (as I used in playground op)

This would require making the `Balance` type parameter a `frame_support::traits::Saturating` everywhere to make the code compile while the only thing we want here is to implement the `Zero` trait. Left unchecked edition for the meantime - as I said, this function is never used in the code anyway.